resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -830,33 +1008,57 @@ void add(DnsRecord r) {
             // we drop servers that have a smaller dots count.
             if (head == null || head.dots < dots) {
                 count = 1;
-                head = new AuthoritativeNameServer(dots, recordName, domainName);
+                head = new AuthoritativeNameServer(dots, r.timeToLive(), recordName, domainName);
             } else if (head.dots == dots) {
                 AuthoritativeNameServer serverName = head;
                 while (serverName.next != null) {
                     serverName = serverName.next;
                 }
-                serverName.next = new AuthoritativeNameServer(dots, recordName, domainName);
+                serverName.next = new AuthoritativeNameServer(dots, r.timeToLive(), recordName, domainName);
                 count++;
             }
         }
 
-        // Just walk the linked-list and mark the entry as removed when matched, so next lookup will need to process
-        // one node less.
-        AuthoritativeNameServer remove(String nsName) {
+        /**
+         * Return the {@link AuthoritativeNameServer} for the {@code nsName} and mark it as handled.
+         */
+        AuthoritativeNameServer handle(String nsName) {
+            // Just walk the linked-list and mark the entry as handled when matched.
+            AuthoritativeNameServer serverName = head;
+
+            while (serverName != null) {
+                if (serverName.nsName.equalsIgnoreCase(nsName)) {
```

Comment:
should this only "handle" the node if it has not yet been handled? For example what about and `ADDITIONAL` section where the same name server name is mapped to multiple IP addresses types.

```
;; ADDITIONAL SECTION:
ns2.google.com.		172800	IN	AAAA	2001:4860:4802:34::a
ns2.google.com.		172800	IN	A	216.239.34.10
```

Do we have a test for this?

```java
if (!serverName.handled && serverName.nsName.equalsIgnoreCase(nsName)) {
..
}
```

Reply:
No we should actually handle both as otherwise one address will be missing in the result. This was a bug we had before (there is a test for this as well now).

Reply:
I agree that we should handle both. This code looks like it will just stop at the first element with a matching name regardless if it is handled, and the subsequent element will never be marked "handled". So it seems like the state of this collection won't be correct in the presence of multiple entries with the same name.

Reply:
ah got it... So you say we should always walk the whole list as there may be more then one that matches ?

Reply:
I misunderstood how this was being used ... so disregard the original suggestion. However it is possible there will be duplicate NS (with different domain) in the authority section that would be problematic (not sure if this is possible).

Reply:
I think this could not happen. So it should be fine ?